### ISSUE TITLE
Investigate ODIS signer fallback mechanism when Azure FrontDoor has degraded service

### ISSUE THREAD

**Issue by kevjue** [celo-org/celo-monorepo#8282]
The Azure Frontdoor service recently had degraded service that significantly impacted the ODIS service, as described in this [incident report](https://www.notion.so/clabsco/ODIS-signer-errors-58e65837162e4477bd68350737654dea#1877f30edbbe44e2865796a0944323d5).

The FrontDoor service is used for DOS protection and SSL termination.  However, in the event of a FD downtime, there should be a fallback.  E.g. perhaps the combiner can directly communicate with the signer clusters.

**Comment by emilhotkowski** [comment]
I know this is not my issue but I am leaving here my fair share of thoughts.

1. We are using k8s cluster with ingress with public IP and over this public IP we have Azure Front Door. This doesn't really makes sense to me, since we are not utilizing "global acceleration" of Azure Front Door but rather we are using this as a DoS protection and SSL termination. Our cluster already has public facing IP that can be DDoS'ed if nginx-ingress is not properly set up and people might not set up any quotas on Ingress since they think we have DoS protection on FD.
2. If we care about latency and performance, check out this [article](https://azure.microsoft.com/pl-pl/blog/application-gateway-ingress-controller-for-azure-kubernetes-service/). This approach seems faster, safer and Application Gateway (even though it has little worse SLA) seems to be a better working service. Check outages https://azure.microsoft.com/pl-pl/blog/application-gateway-ingress-controller-for-azure-kubernetes-service/

@codyborn 

EDIT: Lesson learnt! We have global signers, so we are using global acceleration of Front Door. I'll look into fallback solution instead of changing Front Door since it is the best service for our case (even if faulty recently).

**Comment by emilhotkowski** [comment]
After extra knowledge I gained during last conversation I would like to propose a following path to make system more resilient 
1.  We should make sure that our ingress is DoS protected itself. [LINK](https://www.altostack.io/company/blog/kubernetes-best-practices-security/)
> a well designed Kubernetes ingress policy can add a second layer of protection.
2. Azure offers for global traffic two services "Front Door" and "Traffic Manager", the former fails us from time to time and the latter does not have capability of SSL Termination. That's why I would propose to use **Application Gateway**. We should point with Azure AG to our ingresses and in round-robin fashion, in case of FD failure, we should call signers.
3. To make it easier on ourselves we should consider using Circuit Breaker pattern, possibly with some implementation like https://www.npmjs.com/package/opossum (which is developed by RedHat)
4. My only concern is security. Should be make this Azure AG private with authentication? In a way, ingresses are already public, so it might not be a concern.

@codyborn @kevjue

**Comment by codyborn** [comment]
We currently use the K8s ingress controller to do local load balancing, so no round-robin is needed at the proxy/load-balancer level. In this case, I'm not sure that adding Application Gateway adds anything that FrontDoor doesn't already have.

It's a bit tricky to add authentication since these endpoints need to be publicly available (anyone can use ODIS protocol). 

A simple first step could be to enhance K8s ingress policies (as you mentioned above) and add a fallback to the combiner to bypass the FrontDoor services. We will just need to verify that we can support a non-HTTPS connection for fallback or see if we can easily manage a cert within each K8s cluster.

**Comment by emilhotkowski** [comment]
This is why i wanted to have AG between Combiner and Signers, for ssl termination but also in case of Front Door grant failure we would have a single point of recovery, which would be this AG instead of managing cluster Access on application level.

I'll look into cert management within cluster to verify your version

**Comment by emilhotkowski** [comment]
@codyborn after today's talk we decided that we should stick with simple fallback by using cluster's IP relative to the faulty FrontDoor. In another task it would be great to create an extra layer of DDoS protection on our ingress.

In this task I'll implement Circuit Break pattern to fallback into using IP instead of FD url.

**Comment by emilhotkowski** [comment]
@codyborn once you feel it is done, feel free to close this. I'd be cool if you found time to check this.
